### ISSUE TITLE
Target("wasm") performance issue

### ISSUE THREAD

**Issue by lambneo** [open-policy-agent/opa#3715]
Hi, I tried the advices given in [issue#3707](https://github.com/open-policy-agent/opa/issues/3707#issuecomment-893232730) but to find a performance issue. If I add `rego.Target("wasm)` in `rego.New()`, the time cost will surge to 140ms per Eval(). However the Eval() without it costs only 4ms per time.

## Expected Behavior 
Running time for 100 serial executions should cost 40ms or even less.
## Actual Behavior
Running time for 100 serial executions cost 12.65s.
<img width="289" alt="截屏2021-08-09 上午11 38 48" src="https://user-images.githubusercontent.com/37793830/128658065-74c86d9d-5dc7-45fd-a54e-5c6e3aa3264e.png">

## Steps to Reproduce the Problem
The func codes:
```
	r := rego.New(
		rego.Query(defaultQuery),
		// rego.Target("wasm"),
		rego.Module(PolicyFileName, string(policy)),
		)
       // r.Eval() ...
}
```
And I write a test case to check it:
```
func TestQuery(t *testing.T) {
	for i := 0;i < 100; i++ {
		status, err :=  Eval(input)
		if err != nil {
			t.Fatalf("Unexpected error while compiling query: %v", err)
		}

		if !reflect.DeepEqual(expect, status) {
			t.Fatalf("Expected %v while getitng %v", expect, status)
		}
	}
}
```
The running time for this instance is 0.04s
<img width="303" alt="截屏2021-08-09 下午12 56 11" src="https://user-images.githubusercontent.com/37793830/128661806-a2c3c127-f98e-4324-a7bb-2cf955f25649.png">
 if I make `rego.Target()` work, the running time would be 12.65s
<img width="289" alt="截屏2021-08-09 上午11 38 48" src="https://user-images.githubusercontent.com/37793830/128661818-3ab9ded9-5491-425b-b34c-b2f55aeace46.png">
It seems this option in rego greatly slows down the program. However, the wasm is supposed to be the fastest way to use opa.
Any advice if I want to use wasm-opa for now?

**Comment by srenatus** [comment]
Please look into splitting your eval path -- use `PrepareForEval` first, then `Eval` on int, and re-use the `PrepareForEval` result for further cals to `Eval`.

**Comment by srenatus** [comment]
Also please ensure you're using a recent enough release. We've had a few fixes related to the wasm target in 0.31.0.

**Comment by lambneo** [comment]
Ok got it, Thank you again!
